virttest/libvirt_xml/vm_xml.py

```diff
@@ -3830,3 +3836,77 @@ def __init__(self, virsh_instance=base.virsh):
                                  tag_name='gid')
         super(VMIDMapXML, self).__init__(virsh_instance=virsh_instance)
         self.xml = '<idmap/>'
+
+
+# Sub-element of OS XML
+class VMOSFWXML(base.LibvirtXMLBase):
+    """
+    Firmware tag XML class of OS tag
+
+    Elements:
+        feature: list of dict - enabled, name
+    Example:
+        <firmware>
+        <feature enabled='yes' name='enrolled-keys'/>
+        <feature enabled='yes' name='secure-boot'/>
+        </firmware>
+    """
+
+    __slots__ = ('feature',)
+
+    def __init__(self, virsh_instance=base.virsh):
+        """
+        Create new VMOSFWXML instance
+        """
+        accessors.XMLElementList("feature", self, parent_xpath='/',
+                                 marshal_from=self.marshal_from_feature,
+                                 marshal_to=self.marshal_to_feature)
+        super(VMOSFWXML, self).__init__(virsh_instance=virsh_instance)
+        self.xml = '<firmware/>'
+
+    # Sub-element of firmware
+    class FeatureXML(VMXML):
+        """Feature element of firmware"""
+
+        __slots__ = ('enabled', 'name')
+
+        def __init__(self, virsh_instance=base.virsh):
+            """
+            Create new FeatureXML instance
+            """
+            accessors.XMLAttribute("enabled", self, parent_xpath='/firmware',
+                                   tag_name='feature', attribute='enabled')
+            accessors.XMLAttribute("name", self, parent_xpath='/firmware',
+                                   tag_name='feature', attribute='name')
+            super(VMOSFWXML.FeatureXML, self).__init__(virsh_instance=virsh_instance)
+
+        def update(self, attr_dict):
+            for attr, value in list(attr_dict.items()):
+                setattr(self, attr, value)
+
+    @staticmethod
+    def marshal_from_feature(item, index, libvirtxml):
+        """
+        Convert a FeatureXML instance into tag and attributes of firmware
+        """
+        del index
+        del libvirtxml
+        feature = item.xmltreefile.find("/firmware/feature")
+        try:
+            return(feature.tag, dict(list(feature.items())))
+        except AttributeError:
+            raise xcepts.LibvirtXMLError("Expected a list of feature "
+                                         "instance, not a %s" % str(item))
+
+    @staticmethod
+    def marshal_to_feature(tag, attr_dict, index, libvirtxml):
+        """
+        Convert a tag and attributes to a FeatureXML instance
+        """
+        del index
+        if tag == 'feature':
+            newone = VMOSFWXML.FeatureXML(virsh_instance=libvirtxml.virsh)
+            newone.update(attr_dict)
+            return newone
+        else:
+            return None
```

Comment:
I don't think the definition of the marshal function is correct. Please contact me to have a discussion. I think we have a scheme of this kind.